### PR TITLE
Adding IdentityToken header in readRows call

### PR DESCRIPTION
**Pull request by isha97**

### CODE REVIEW

File: bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/IdentityTokenSupplier.java [hunk 1]
```diff
@@ -23,73 +22,41 @@ public class IdentityTokenSupplier implements Serializable {
...
-    cachedToken = null;
-    lastFetchTime = 0;
-  }
-
-  @Override
-  public int hashCode() {
-    return Objects.hashCode(cachedToken, lastFetchTime);
-  }
+  public static Optional<String> createIdentityToken(String audience) {
+    CloseableHttpClient httpClient = HttpClients.custom().setDefaultRequestConfig(config).build();
```
Comment (vishalkarve15) [inline bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/IdentityTokenSupplier.java:35]: Have we looked into https://cloud.google.com/docs/authentication/get-id-token#java?

File: bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/StreamCombiningIterator.java [hunk 2]
```diff
@@ -201,10 +215,31 @@ private void completeStream(boolean addEos) {
     }
   }
 
+  private String parseReadSessionId(String stream) {
+    int streamsIndex = stream.indexOf("/streams");
+    return streamsIndex != -1 ? stream.substring(0, streamsIndex) : stream;
```
Comment (vishalkarve15) [inline bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/StreamCombiningIterator.java:220]: We should have info logs if the readStream string does not contain "/streams"

File: bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/StreamCombiningIterator.java [hunk 3]
```diff
@@ -57,6 +66,11 @@ public class StreamCombiningIterator implements Iterator<ReadRowsResponse> {
   Object last;
   volatile boolean cancelled = false;
   private final Collection<Observer> observers;
+  private static final long IDENTITY_TOKEN_TTL_IN_MILLIS = 50 * 60 * 1000;
+  private static final LoadingCache<String, Optional<String>> READ_SESSION_TO_IDENTITY_TOKEN_CACHE =
+      CacheBuilder.newBuilder()
+          .expireAfterWrite(IDENTITY_TOKEN_TTL_IN_MILLIS, TimeUnit.MILLISECONDS)
```
Comment (davidrabinowitz) [inline bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/StreamCombiningIterator.java:72]: Notice that `.expireAfterWrite(IDENTITY_TOKEN_TTL_IN_MINUTES, TimeUnit.MINUTES)` is more readable

**Review by davidrabinowitz — APPROVED** [review]

File: bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/IdentityTokenSupplier.java [hunk 4]
```diff
@@ -23,73 +22,41 @@ public class IdentityTokenSupplier implements Serializable {
...
-        cachedToken = null;
-      } finally {
-        try {
-          Closeables.close(httpClient, true);
-        } catch (IOException e) {
-          // nothing to do
-        }
+        Closeables.close(httpClient, true);
+      } catch (IOException e) {
+        // nothing to do
```
Comment (shailesh-tripathi-1995) [inline bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/IdentityTokenSupplier.java:56]: Should we add logs for this case well?

File: bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/StreamCombiningIterator.java (hunk 3 above)
Comment (isha97) [reply bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/StreamCombiningIterator.java]: Fixed

File: bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/IdentityTokenSupplier.java (hunk 4 above)
Comment (isha97) [reply bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/IdentityTokenSupplier.java]: Updated the fetch logic to use IdTokenProvider library

File: bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/StreamCombiningIterator.java (hunk 2 above)
Comment (isha97) [reply bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/StreamCombiningIterator.java]: Done

**Pull request merged**